### DISCUSSION TITLE
Basic client example isn't working - any ideas on how to proceed with 415 error code?

### DISCUSSION THREAD

**Discussion by jeffnsquared** [connectrpc/connect-es#622] [Q&A] [answered]
Hello, I have a Kotlin-backed gRPC endpoint which is verified to work with different clients in Kotlin/Java. I'm attempting to access it through the protobuf-es + connect-es setup in my Node TS application.

I am receiving the following client side error 
```Found error: ConnectError:2. Details  with Message [unknown] HTTP 415```

The server side error isn't really helpful
```415 "POST /api/MockSearch HTTP/2" - "-" 2 "Unknown" "@bufbuild/connect-web" "679e2dc4-24c8-4813-b0f9-f87d80818d93" "[HOST]"```

I've tried 3 different transport options (GrpcTransport, GrpcWebTransport, and ConnectTransport) but they all fail with a simple gRPC request using the example PromiseClient.

Example transport
```
		const grpcTransport = createGrpcTransport({
			// Requests will be made to <baseUrl>/<package>.<service>/method
			baseUrl: URL,

			// You have to tell the Node.js http API which HTTP version to use.
			httpVersion: "2",

			// Interceptors apply to all calls running through this transport.
			interceptors: [],
			useBinaryFormat: true
		});
```

Any ideas on how to proceed?

I've used https://github.com/stephenh/ts-proto + https://github.com/deeplay-io/nice-grpc and am able to access the gRPC endpoint so its either something with the generated TS files (I doubt it) or the client call itself.

**Comment by NfNitLoop** [comment]
> Kotlin-backed gRPC endpoint

> I've tried 3 different transport options

Which protocol is your kotlin-backed gRPC endpoint serving?  I'd guess "normal" gRPC, so only grpcTransport will work.  I assume you've tried with `useBinaryFormat: false` too?

An [HTTP 415] error means that the client sent/requested an "Unsupported Media Type", which I'd expect if you're trying to connect with connectTransport or grpcWebTransport if your server isn't actually serving those protocols.

You show how you configure the transport. Can you show how you import and configure the client, too?  You're using connect-node, right? <https://www.npmjs.com/package/@bufbuild/connect-node>

[HTTP 415]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/415

**Reply by jeffnsquared** [reply]
Hello, I tried `useBinaryFormat` with both false and true. You are correct - my kotlin-backed gRPC endpoint is serving "normal" gRPC.

for the client, I did the following:
```
import { createGrpcTransport } from "@bufbuild/connect-node";
import { createPromiseClient } from "@bufbuild/connect";

		const grpcTransport = createGrpcTransport({
			// Requests will be made to <baseUrl>/<package>.<service>/method
			baseUrl: SEARCH_URL,

			// You have to tell the Node.js http API which HTTP version to use.
			httpVersion: "1.1",

			// Interceptors apply to all calls running through this transport.
			interceptors: [],
		});

                const client = createPromiseClient(CustomerService, grpcTransport);
                const mockSearchRequest: MockSearchRequest = new MockSearchRequest({
			searchText: q
		});
                const res = await client.mockSearch(mockSearchRequest);
		return res;
```

any ideas?

**Reply by timostamm** [reply]
Hey Jeff, with the setup you posted, we send a request with `content-type: application/grpc+proto`. It does seem odd that you get a HTTP 415 response. It might be an incorrect URL path. 

I would be surprised if the Kotlin service were not expecting HTTP/2. I would definitely go with `httpVersion: "2"` in the client. Is it also using TLS, or is it using cleartext? For TLS, make sure that the `baseUrl` starts with the `https://` scheme. For cleartext, use plain `http://`. 

The `POST /api/MockSearch` in your server log seems strange. If your service is named `api.CustomerService`, I would expect a `POST /api.CustomerService/MockSearch`. For a gRCP server, you typically want a `baseUrl` like `https://demo.connect.build` in the client.

**Reply by jeffnsquared** [reply]
Hey Timo, for the other requests, its based on HTTP/2 so you are correct. As for the logs, I edited it a bit because I wanted to simplify the logs. But you are correct, the endpoint is indeed `POST /api.CustomerService/MockSearch`. 

I believe I've found the issue. My Kotlin service is only accepting `application/grpc`, but not allowing `application/grpc+proto`. I believe I need to fix this on the server side. But out of curiosity, is there any way I can set the content type in the application code using the connect-es library?

**Reply by timostamm** [reply, marked as answer]
The [spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests) allows both `application/grpc` and `application/grpc+proto` for the binary protobuf format, so it might be a good call to allow both on the server.

You can use an interceptor to modify the request headers:

```ts
const modifyProtocolHeaders: Interceptor = (next) => async (req) => {
  if (req.header.get("Content-Type") === "application/grpc+proto") {
    req.header.set("Content-Type", "application/grpc");
  }
  return await next(req);
}
```

Usage:

```ts
createConnectTransport({
  baseUrl: "https://example.com",
  httpVersion: "2",
  interceptors: [ modifyProtocolHeaders ]
});
```

**Reply by jeffnsquared** [reply]
Thank you @timostamm - this helps as a temporary workaround. I will make the adjustment on server too. Cheers
